### DISCUSSION TITLE
Step3/Step4 NBRA Workflow with CP2K/xTB

### DISCUSSION THREAD

**Discussion by r2stanton** [Quantum-Dynamics-Hub/libra-code#170] [Q&A] [answered]
I have a couple quick questions about step4 in the nbra workflow in Libra. Specifically I'm following the notebook here pretty closely https://github.com/compchem-cybertraining/Tutorials_Libra/tree/master/6_dynamics/2_nbra_workflows/9_step4_cp2k.

First I noticed there's 'init_state' and 'istate' variables in run_dynamics in the first code block, and that init_state appears to take precedence when looking at initial populations in the SH_pop files. Is istate doing something else entirely, or do these serve pretty much the same purpose?

Then also there appears to be some state tracking in step4 that you can set in params_common. Is this only necessary if doing the dynamics in the KS basis? Because it looked like state tracking was also done in step3 for the mixed basis.

**Comment by alexvakimov** [comment, marked as answer]
Hi Robert @r2stanton ,

If you are talking about the `init_states` variable (list) present at the top of the `run_dynamics` function - then yes - it does take precedence over `init_state` listed in the dictionary of the dynamical parameters. In this case, `init_state` is essentially a dummy variable and may be omitted altogether. What happens is that when the `step4.namd_workflow` functions is called it will run the `run_tsh` functions with various sets of variables (methods, initial conditions, batches) in parallel (to some extent). The actual parameters for those runs would be determined by what you list in `init_states`, and other variables in the beginning of the `run_dynamics` function. 

About the state tracking - yes, if you have done it in step3, you won't need to do it in step4. Just the heads up: this whole thing about state tracking at different steps will be removed in the new releases to make it more intuitive.

**Reply by r2stanton** [reply]
@alexvakimov 
Hi Professor Akimov,

Great thank you that definitely clears many things up. Are there any rough estimates for how long these processes should be taking, as I might be running into a bottleneck somewhere, but I'm not sure if it's just because large systems the xTB approach can deal with have so many states.

To give some specific numbers, I have ~130*2 states, and it takes ~20 minutes for a batch of 10 trajectories across 1000 timesteps to complete on a single process. Is that to be expected simply from the large system size, or does that sound high? At ~2min per trajectory, sampling different excitations, SH methods, and initial conditions with a reasonable number of trajectories is seeming like it will get a bit expensive quickly. Either way, thanks in advance for any info about that!

**Reply by r2stanton** [reply]
Just as an update to the above comment in case anyone checks this in the future:

If doing RKS for the dynamics, actually chopping only one of the spin channels out of the E, Hvib, S, St files in the res-ks directory can help quite a bit with reducing not only the time taken mentioned above for the NAMD, but also the memory reqs associated with loading in the necessary files which helped me out a good bit with a slightly different parallelization scheme. I believe for the mixed basis, a similar thing should work as well by pulling out a relevant block diagonal, but I haven't verified this.
